addons/mail/static/src/messaging/entity/attachment/attachment.js

```diff
@@ -157,126 +210,131 @@ function AttachmentFactory({ Entity }) {
         }
 
-        /**
-         * Unlink the provided attachment.
-         */
-        async remove() {
-            await this.env.rpc({
-                model: 'ir.attachment',
-                method: 'unlink',
-                args: [this.id],
-            }, { shadow: true });
-            this.delete();
-        }
-
-        //----------------------------------------------------------------------
-        // Private
-        //----------------------------------------------------------------------
-
         /**
          * @override
          */
         _createInstanceLocalId(data) {
             const { id, isTemporary = false } = data;
             if (isTemporary) {
-                return `${this.constructor.name}_${nextTemporaryId}`;
+                return `${this.constructor.entityName}_${nextTemporaryId}`;
             }
-            return `${this.constructor.name}_${id}`;
+            return `${this.constructor.entityName}_${id}`;
         }
 
         /**
          * @override
          */
-        _update(data) {
-            let {
-                composers,
-                filename,
-                id = getAttachmentNextTemporaryId(),
-                isTemporary = false,
-                mimetype = '',
-                name,
-                res_id,
-                res_model,
-                size,
-            } = data;
-
-            Object.assign(this, {
-                filename,
-                id,
-                isTemporary,
-                mimetype,
-                name,
-                size,
-            });
-
-            // composers
-            if (composers !== undefined) {
-                if (composers === null) {
-                    this.unlink({ composers: null });
-                } else {
-                    const prevComposers = this.composers;
-                    const prevOldComposers = prevComposers.filter(composer => !composers.includes(composer));
-                    const newComposers = composers.filter(composer => !prevComposers.includes(composer));
-                    this.unlink({ composers: prevOldComposers });
-                    this.link({ composers: newComposers });
-                }
-            }
-            // originThread
-            if (res_id && res_model) {
-                let newOriginThread = this.env.entities.Thread.fromModelAndId({
-                    id: res_id,
-                    model: res_model,
-                });
-                if (!newOriginThread) {
-                    newOriginThread = this.env.entities.Thread.create({
-                        id: res_id,
-                        model: res_model,
-                    });
-                }
-                const prevOriginThread = this.originThread;
-                if (newOriginThread !== prevOriginThread) {
-                    this.link({ originThread: newOriginThread });
-                }
-            }
+        _updateAfter(previous) {
```

Comment:
Don't really understand the issue, but I think the confusion comes from `_update` being renamed to `_updateAfter` in this PR. I decided to incorporate previous `_update()` in the new update cycle to help doing the conversion step by step and also helps keeping remaining code in `_update` that are still hard to convert.

Now the reason for its rename is because old `_update()` liked to make changes based on previous data. This was nice when `_update()` was just by itself, but now that there are automatic updates (directs & computes), we have to keep track of previous state and then invoke old `_update()` (and after the automatic updates, because it may require results of these auto-updates). So there a method `_updateBefore()` to define which previous data to track, and when `_updateAfter()` is called, it has previous data as one of its parameter.